src/session/content/message_row/mod.rs

```diff
@@ -2,149 +2,320 @@ mod sticker;
+                    content_natural_baseline,
+                ) = {
+                    let (minimum, mut natural, minimum_baseline, natural_baseline) =
+                        content.measure(orientation, remaining_for_size);
+                    let (_, default_natural, _, _) = content.measure(orientation, -1);
+
+                    // Always prefer default natural size
+                    if natural > default_natural && default_natural >= minimum {
+                        natural = default_natural
+                    }
+
+                    (minimum, natural, minimum_baseline, natural_baseline)
+                };
+
+                minimum_baseline = content_minimum_baseline;
+                natural_baseline = content_natural_baseline;
+
+                if let gtk::Orientation::Horizontal = orientation {
+                    minimum += content_minimum;
+                    natural += content_natural;
+                } else {
+                    minimum = minimum.max(content_minimum);
+                    natural = natural.max(content_natural);
+                }
+            }
+
+            (minimum, natural, minimum_baseline, natural_baseline)
+        }
+
+        fn size_allocate(&self, _widget: &Self::Type, width: i32, height: i32, baseline: i32) {
+            let mut remaining_width = width;
+
+            if let Some(avatar) = self.avatar.borrow().as_ref() {
+                let (_, natural_size) = avatar.preferred_size();
+                let allocation = gtk::Allocation {
+                    x: 0,
+                    y: height - natural_size.height,
+                    width: natural_size.width,
+                    height: natural_size.height,
+                };
+
+                avatar.size_allocate(&allocation, -1);
+
+                remaining_width -= natural_size.width + SPACING;
+            }
+
+            if let Some(content) = self.content.borrow().as_ref() {
+                let (_, natural_size) = content.preferred_size();
+                let actual_width = remaining_width.min(natural_size.width);
+                let x = if self.is_outgoing.get() {
+                    width - actual_width
+                } else {
+                    width - remaining_width
+                };
+
+                let allocation = gtk::Allocation {
+                    x,
+                    y: 0,
+                    width: actual_width,
+                    height,
+                };
+                content.size_allocate(&allocation, baseline);
+            }
+        }
+
+        fn request_mode(&self, _widget: &Self::Type) -> gtk::SizeRequestMode {
+            gtk::SizeRequestMode::HeightForWidth
+        }
+    }
 }
 
 glib::wrapper! {
     pub struct MessageRow(ObjectSubclass<imp::MessageRow>)
         @extends gtk::Widget;
 }
 
-impl Default for MessageRow {
-    fn default() -> Self {
-        Self::new()
+impl MessageRow {
+    pub fn new(message: &Message) -> Self {
```

Comment:
This is actually not used anywhere In the code.
I plan to change the visibility modifiers to `pub(crate)` and `pub(super)` soon everywhere in the code, if you don't mind.
Then clippy can warn us about such unused code.

Reply:
> I plan to change the visibility modifiers to pub(crate)

That seems a good idea.